<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-rafaelsangarcia/blob/main/Actividad3_A01182780.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#ACTIVIDAD 3
**RAFAEL SANCHEZ GARCIA**


#PART 1

**Fundamentos de bases de datos y ciencia de datos**
200 palabras

#PART 2

## LOAD DATA

In [79]:
#LOAD THE DATABASE TO MY COLAB
input = "https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv"

In [80]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
df = pd.read_csv(input)
df.head()

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15      X16  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0      0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   3455.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0  14948.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0  28959.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0  19146.0   

       X17     X18      X19      X20     X21     X22     X23    Y  
0      0.0     0.0    689.0      0.0     0.0     0.0     0.0  1.0  
1   3261.0     0.0   1000.0   1000.0  1000.0     0.0  2000.0  1.0  
2  15549.0  1518.0   1500.0   1000.0  1000.0  1000.0  5000.0  0.0  
3  29547.0  2000.0   2019.0   1200.0  1100.0  1069.0  1000.0  0.0  
4  19131.0  2000.0  36681.0  10000.0  9000.0   689.0   679.0  0.0  

[5 rows x 25 columns]

**Lets verify is we have NAs**

In [81]:
df.isnull().values.any()

True

**We have Null values, so we need to start to decide what to do which them. But first lets explore which columns have nulls**

In [82]:
df.isnull().any()

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

**All except ID and X1 has nulls**

## CLEAN DATA

**Lets analysis the data**


In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

**ID and X1 are int, the rest are float,Lets see how are the distribution**

In [84]:
#better way to see the data
df.describe().transpose()

count           mean            std       min       25%       50%  \
ID   30000.0   15000.500000    8660.398374       1.0   7500.75   15000.5   
X1   30000.0  167484.322667  129747.661567   10000.0  50000.00  140000.0   
X2   29999.0       1.603753       0.489125       1.0      1.00       2.0   
X3   29998.0       1.853057       0.790320       0.0      1.00       2.0   
X4   29998.0       1.551903       0.521968       0.0      1.00       2.0   
X5   29995.0      35.484214       9.218024      21.0     28.00      34.0   
X6   29997.0      -0.016635       1.123829      -2.0     -1.00       0.0   
X7   29995.0      -0.133689       1.197254      -2.0     -1.00       0.0   
X8   29993.0      -0.166405       1.196048      -2.0     -1.00       0.0   
X9   29991.0      -0.220800       1.169153      -2.0     -1.00       0.0   
X10  29984.0      -0.266342       1.133296      -2.0     -1.00       0.0   
X11  29986.0      -0.291136       1.150134      -2.0     -1.00       0.0   
X12  29989.0   51236.862750   73645.219278 -165580.0   3565.00   22387.0   
X13  29989.0   49190.734669   71183.385123  -69777.0   2986.00   21207.0   
X14  29987.0   47025.350152   69360.863317 -157264.0   2667.50   20089.0   
X15  29985.0   43275.652326   64345.500073 -170000.0   2329.00   19052.0   
X16  29983.0   40324.493980   60809.984983  -81334.0   1763.50   18107.0   
X17  29990.0   38881.135745   59561.312967 -339603.0   1256.25   17081.0   
X18  29992.0    5662.945886   16564.165089       0.0   1000.00    2100.0   
X19  29991.0    5922.488913   23044.177075       0.0    835.50    2009.0   
X20  29992.0    5225.623400   17608.422625       0.0    390.00    1800.0   
X21  29989.0    4827.252526   15668.751975       0.0    296.00    1500.0   
X22  29989.0    4800.297209   15280.842069       0.0    251.00    1500.0   
X23  29995.0    5216.259977   17778.848359       0.0    118.00    1500.0   
Y    29997.0       0.221189       0.415054       0.0      0.00       0.0   

           75%        max  
ID    22500.25    30000.0  
X1   240000.00  1000000.0  
X2        2.00        2.0  
X3        2.00        6.0  
X4        2.00        3.0  
X5       41.00       79.0  
X6        0.00        8.0  
X7        0.00        8.0  
X8        0.00        8.0  
X9        0.00        8.0  
X10       0.00        8.0  
X11       0.00        8.0  
X12   67139.00   964511.0  
X13   64027.00   983931.0  
X14   60182.00  1664089.0  
X15   54560.00   891586.0  
X16   50213.00   927171.0  
X17   49208.25   961664.0  
X18    5006.00   873552.0  
X19    5000.00  1684259.0  
X20    4505.50   896040.0  
X21    4014.00   621000.0  
X22    4033.00   426529.0  
X23    4000.00   528666.0  
Y         0.00        1.0

### Massive drop

**Vamos a eliminar todos los renglones que tengan mas del 10% de la informacion faltante por columnas**

In [85]:
#Calcular al menos cuantas columnas debe de tener con valores
df.shape[1]*.90

22.5

**Vamos a cerrarlo en 22**

In [86]:
ndf = df.copy()
ndf.dropna(thresh=22, inplace = True)
ndf.info()
#ndf.reset_index()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29988 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      29988 non-null  int64  
 1   X1      29988 non-null  int64  
 2   X2      29988 non-null  float64
 3   X3      29988 non-null  float64
 4   X4      29988 non-null  float64
 5   X5      29985 non-null  float64
 6   X6      29988 non-null  float64
 7   X7      29988 non-null  float64
 8   X8      29986 non-null  float64
 9   X9      29984 non-null  float64
 10  X10     29980 non-null  float64
 11  X11     29982 non-null  float64
 12  X12     29988 non-null  float64
 13  X13     29988 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29987 non-null  float64
 18  X18     29988 non-null  float64
 19  X19     29987 non-null  float64
 20  X20     29988 non-null  float64
 21  X21     29985 non-null  float64
 22

In [87]:
#Calcular el porcentaje de informacion que quitamos.
(1-ndf.shape[0]/30000)*100

0.039999999999995595

**Solo se eliminaron 30000 - 29988 = 12 renglones lo cual representa aprox el 0.039 % de la informacion, con lo cual no se vera afectada el muestreo**

### **Lets start with column Y**

In [ ]:
ndf['Y'].isnull().any()

True

We can see that the item 6232, 29824 and 29825 has a lots of NaN, we can eliminate, and it will not affect too much to eliminate because it is just 3 items out 30,000

In [ ]:
ndfIndex = ndf[ndf['Y'].isnull()].index
ndfIndex
#ndf = ndf.drop(6232)
ndf =ndf.drop(ndfIndex);

In [ ]:
ndf[ndf['Y'].isnull()]

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

In [ ]:
ndf['Y'].isnull().any()

False

**Column Y does not have more null values.**

In [88]:
ndf.isnull().any()

ID     False
X1     False
X2     False
X3     False
X4     False
X5      True
X6     False
X7     False
X8      True
X9      True
X10     True
X11     True
X12    False
X13    False
X14     True
X15     True
X16     True
X17     True
X18    False
X19     True
X20    False
X21     True
X22     True
X23     True
Y      False
dtype: bool

### Lets work with column X5

In [89]:
ndf[ndf['X5'].isnull()]

ID      X1   X2   X3   X4  X5   X6   X7   X8   X9  ...       X15  \
182  183  500000  2.0  1.0  1.0 NaN  0.0  0.0  0.0  0.0  ...  122967.0   
219  220  310000  2.0  1.0  2.0 NaN -1.0 -1.0 -1.0 -1.0  ...       0.0   
239  240  140000  2.0  2.0  3.0 NaN  0.0  0.0  0.0  0.0  ...   19068.0   

          X16      X17      X18      X19      X20      X21      X22      X23  \
182  108834.0  70064.0  70010.0  30357.0  30000.0  20000.0  52183.0  20000.0   
219       0.0      0.0   4542.0    126.0      0.0      0.0      0.0      0.0   
239   16409.0  16383.0   3000.0   2000.0   2198.0   1000.0   3000.0   2399.0   

       Y  
182  0.0  
219  0.0  
239  1.0  

[3 rows x 25 columns]

**Vamos a llenar los NaN, con la mediana, ya que la muestra es lo suficientemente grande, no tendremos problemas de sesgo**

In [90]:
x5_median = ndf['X5'].median()
print(x5_median)
ndf['X5'].fillna(value = x5_median, 
                    inplace = True)


34.0


In [91]:
ndf.isnull().any()

ID     False
X1     False
X2     False
X3     False
X4     False
X5     False
X6     False
X7     False
X8      True
X9      True
X10     True
X11     True
X12    False
X13    False
X14     True
X15     True
X16     True
X17     True
X18    False
X19     True
X20    False
X21     True
X22     True
X23     True
Y      False
dtype: bool

In [161]:
ndf[ndf['X8'].isnull()].transpose()

17989    18003
ID   17990.0  18004.0
X1   80000.0  70000.0
X2       1.0      1.0
X3       2.0      2.0
X4       1.0      2.0
X5      46.0     33.0
X6       0.0      1.0
X7       0.0      2.0
X8       NaN      NaN
X9       0.0      7.0
X10      0.0      7.0
X11      0.0      6.0
X12  79707.0  26181.0
X13  79876.0  37611.0
X14  79919.0  37026.0
X15  64237.0  38922.0
X16  47890.0  38318.0
X17  48036.0  38133.0
X18   3542.0  12000.0
X19   3500.0      0.0
X20   2500.0   2500.0
X21   2000.0      0.0
X22   2500.0   1000.0
X23   2000.0   1500.0
Y        0.0      0.0

In [94]:
ndf3 = ndf.copy()
ndf3 = ndf3[['ID','X8','X9','X10','X11']]
ndf3[ndf3['ID']==18004].mode(axis = 1)[0]

18003    7.0
Name: 0, dtype: float64

In [96]:
ndf3[ndf3['X8'].isnull()]

ID  X8   X9  X10  X11
17989  17990 NaN  0.0  0.0  0.0
18003  18004 NaN  7.0  7.0  6.0

In [138]:
ndf3[ndf3['ID']==18004]

ID   X8   X9  X10  X11
18003  18004  9.0  7.0  7.0  6.0

In [ ]:
listColumns=['X8','X9','X10','X11']
#listColumns=[1,2,3,4]
listItem=[18004]
moda_list= []
for j in listItem:
  moda_list= []
  
  for i in listColumns:
    moda_list.append(ndf3[ndf3['ID']==j].iloc[0,i])

  m = pd.Series(moda_list).mode()[0]
  ndf3.at[18003,'X8']=m
    

In [176]:
ndf3.loc[17989][1]
ndf3=ndf.copy()

In [174]:
listColumns=['X8','X9','X10','X11']
#listColumns=[1,2,3,4]
listItem=[18003,17989]
moda_list= []
for j in listItem:
  moda_list= []
  
  for i in listColumns:
    moda_list.append(ndf3.loc[j][i])

  m = pd.Series(moda_list).mode()[0]
  ndf3.at[j,'X8']=m
    

In [175]:
ndf3[ndf3['X8'].isnull()]
ndf3.loc[17989]


ID     17990.0
X1     80000.0
X2         1.0
X3         2.0
X4         1.0
X5        46.0
X6         0.0
X7         0.0
X8         0.0
X9         0.0
X10        0.0
X11        0.0
X12    79707.0
X13    79876.0
X14    79919.0
X15    64237.0
X16    47890.0
X17    48036.0
X18     3542.0
X19     3500.0
X20     2500.0
X21     2000.0
X22     2500.0
X23     2000.0
Y          0.0
Name: 17989, dtype: float64

In [ ]:
ndf4=ndf.copy()
ndf4.fillna(ndf4['X8'].mode()[0],inplace= True,axis =0)
ndf4[ndf4['ID']==18004].iloc[0,8]

0.0

In [259]:
from pandas.core.algorithms import mode
ndf5=ndf.copy()
col = ndf5.loc[: , "X8":"X11"]
ndf5['rafa'] = col.mean(axis=1)
col.mode(axis=1)


0    1   2   3
0     -2.0 -1.0 NaN NaN
1      0.0  NaN NaN NaN
2      0.0  NaN NaN NaN
3      0.0  NaN NaN NaN
4      0.0  NaN NaN NaN
...    ...  ...  ..  ..
29995  0.0  NaN NaN NaN
29996 -1.0  0.0 NaN NaN
29997  0.0  NaN NaN NaN
29998  0.0  NaN NaN NaN
29999  0.0  NaN NaN NaN

[29988 rows x 4 columns]